### PR TITLE
Added solution for Euler number 12.

### PR DESCRIPTION
**Pull request by jorander**
I'm doing this to learn FP and Javaslang, so all kind of comments are highly appreciated.

BR,
Jörgen

### CODE REVIEW

**Comment by ashrko619** [comment]
Hi @jorander ,
Your code works fine :+1: But there is a much faster algorithm based on the fact that n, n+1 are co-primes. See this http://codereview.stackexchange.com/a/83680  :)

**Comment by jorander** [comment]
Hi @ashrko619 ,
Thanks for pointing that out. If I understand the answer you linked correctly it is the calculation of number of divisors that could be smarter. That would mean calculating the number of divisors directly, taking advantage of some more involved mathematical idiom, rather than first creating a stream of factors and then count them.

I'm sure it will be faster, but I'm not sure that is the right path to go for the Euler problems as tests for Javaslang. I think the main objective here is to validate and show case Javaslang's capabilities as an API for functional programming in Java. In doing so I would like to keep the logic as simple as possible and the solutions as close as possible to the problem statement so that any reader should be able to focus on the Javaslang constructs rather than pondering the mathematical logic we use to arrive at the right solution. From that perspective I think the more naïve solution I used is preferable.

Other point-of-views? //cc @danieldietrich 

Regards,
Jörgen

File: javaslang/src/test/java/javaslang/collection/euler/Euler12Test.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+     * What is the value of the first triangle number to have over five hundred
+     * divisors?
+     */
+    @Test
+    public void shouldSolveProblem12() {
+        assertThat(valueOfFirstTriangleNumberWithMoreDivisorsThan(5)).isEqualTo(28);
+        assertThat(valueOfFirstTriangleNumberWithMoreDivisorsThan(500)).isEqualTo(76_576_500);
+    }
+
+    private static long valueOfFirstTriangleNumberWithMoreDivisorsThan(long nofDivisors) {
```
Comment (paplorinc) [inline javaslang/src/test/java/javaslang/collection/euler/Euler12Test.java]: `nof`?
I think `divisorCount` would make more sense?

File: javaslang/src/test/java/javaslang/collection/euler/Euler12Test.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+     * We can see that 28 is the first triangle number to have over five
+     * divisors.
+     * <p>
+     * What is the value of the first triangle number to have over five hundred
+     * divisors?
+     */
+    @Test
+    public void shouldSolveProblem12() {
+        assertThat(valueOfFirstTriangleNumberWithMoreDivisorsThan(5)).isEqualTo(28);
+        assertThat(valueOfFirstTriangleNumberWithMoreDivisorsThan(500)).isEqualTo(76_576_500);
```
Comment (paplorinc) [inline javaslang/src/test/java/javaslang/collection/euler/Euler12Test.java]: how did you come up with the end result :)?

File: javaslang/src/test/java/javaslang/collection/euler/Utils.java [hunk 3]
```diff
@@ -27,6 +27,13 @@ public static BigInteger factorial(int n) {
         return Stream.rangeClosed(1, n).map(BigInteger::valueOf).fold(BigInteger.ONE, BigInteger::multiply);
     }
 
+    public static Stream<Long> factors(long l) {
```
Comment (paplorinc) [inline javaslang/src/test/java/javaslang/collection/euler/Utils.java]: please rename `l`, it looks like `1` :)
```java
factors(long number)
```

File: javaslang/src/test/java/javaslang/collection/euler/Euler12Test.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        return triangleNumbers()
+                .find(t -> nofDivisors(t) > nofDivisors)
+                .get();
+    }
+
+    private static Stream<Long> triangleNumbers() {
+        return Stream.from(1L).scanLeft(0L, (a, l) -> a + l);
+    }
+
+    private static long nofDivisors(long l) {
```
Comment (paplorinc) [inline javaslang/src/test/java/javaslang/collection/euler/Euler12Test.java]: same

File: javaslang/src/test/java/javaslang/collection/euler/Utils.java (hunk 3 above)
Comment (ruslansennov) [inline javaslang/src/test/java/javaslang/collection/euler/Utils.java]: please stop this flood

File: javaslang/src/test/java/javaslang/collection/euler/Euler12Test.java (hunk 2 above)
Comment (jorander) [inline javaslang/src/test/java/javaslang/collection/euler/Euler12Test.java]: You mean that the first triangle number with more than 500 divisors is 76 576 500? I calculated it using my program and then went to https://projecteuler.net/problem=12 to enter my answer and get it evaluated. It turned out to be right.

**Comment by danieldietrich** [comment]
> From that perspective I think the more naïve solution I used is preferable.
> Other point-of-views? //cc @danieldietrich

I also prefer the more naïve solution. My intention of having Eule tests is to implement real world examples with the help of the Javaslang API. The more concise the code is, the better. Often optimization means low level, plain Java code. That's not the goal here.

> Is there an equivalent for distributing the workload for data in a Javaslang Stream over more cores?

I had it in mind but not further investigated it yet. The _work_ has to be splitted into parts and the result have to be joined again (in the right order, if there is an order). Most of the Javaslang types are Iterable, i.e. they have a method `splitterator()`. Javaslang's Stream is a _linear_ sequence, i.e. hard/impossible to split. Java's Stream works in a different way. It is not a Stream as we know it from Scala. In Javaslang maybe we could leverage the Future implementation to move parts of a computation to different threads but I have to think about an API here...

**Comment by danieldietrich** [comment]
Thank you very much!

**Pull request merged**